core/node/crypto/client.go

```diff
@@ -2,140 +2,404 @@ package crypto
 
 import (
 	"context"
+	"encoding/binary"
 	"encoding/hex"
+	"fmt"
 	"math/big"
+	"strings"
 
 	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/ethclient/simulated"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/prometheus/client_golang/prometheus"
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/trace"
+
+	"github.com/river-build/river/core/node/infra"
 )
 
+type simulatedClientWrapper struct {
+	simulated.Client
+}
+
+var _ BlockchainClient = (*simulatedClientWrapper)(nil)
+
+// NewWrappedSimulatedClient returns a wrapped client that implements methods
+// missing on the simulated client which are occasionally used by abi-genned
+// contract code.
+func NewWrappedSimulatedClient(client simulated.Client) BlockchainClient {
+	return &simulatedClientWrapper{
+		Client: client,
+	}
+}
+
+func (scw *simulatedClientWrapper) CallContractAtHash(
+	ctx context.Context,
+	msg ethereum.CallMsg,
+	blockHash common.Hash,
+) ([]byte, error) {
+	bh, ok := scw.Client.(bind.BlockHashContractCaller)
+	if ok {
+		return bh.CallContractAtHash(ctx, msg, blockHash)
+	}
+
+	block, err := scw.BlockByHash(ctx, blockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return scw.CallContract(ctx, msg, block.Number())
+}
+
+func (scw *simulatedClientWrapper) CodeAtHash(
+	ctx context.Context,
+	account common.Address,
+	blockHash common.Hash,
+) ([]byte, error) {
+	bh, ok := scw.Client.(bind.BlockHashContractCaller)
+	if ok {
+		return bh.CodeAtHash(ctx, account, blockHash)
+	}
+
+	block, err := scw.BlockByHash(ctx, blockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return scw.CodeAt(ctx, account, block.Number())
+}
+
 type otelEthClient struct {
 	*ethclient.Client
-	tracer trace.Tracer
+	ethCalls *prometheus.CounterVec
+	tracer   trace.Tracer
+	chainId  string
 }
 
 var _ BlockchainClient = (*otelEthClient)(nil)
 
-// NewInstrumentedEthClient wraps an Ethereum client and adds open-telemetry tracing.
-func NewInstrumentedEthClient(client *ethclient.Client, tracer trace.Tracer) *otelEthClient {
-	return &otelEthClient{Client: client, tracer: tracer}
+// NewInstrumentedEthClient wraps an Ethereum client and adds open-telemetry tracing and metrics
+// collection.
+func NewInstrumentedEthClient(
+	client *ethclient.Client,
+	chainId uint64,
+	metrics infra.MetricsFactory,
+	tracer trace.Tracer,
+) *otelEthClient {
+	var ethCalls *prometheus.CounterVec
+	if metrics != nil {
```

Comment:
in other parts of the code to eliminate these nil checks it's assumed that metrics is always initialized. In test it's empty metrics are initialized like this: `infra.NewMetricsFactory(nil, "", ""),` let's drop nil checks here as well.